oak_functions/loader/src/server.rs

```diff
@@ -249,6 +280,22 @@ impl WasmState {
         .context("failed to instantiate Wasm module")?
         .assert_no_start();
 
+        let alloc = instance
+            .export_by_name(ALLOC_FUNCTION_NAME)
+            .context("could not find Wasm `alloc` export")?
+            .as_func()
+            .cloned()
+            .context("could not interpret Wasm `alloc` export as function")?;
+        let expected_signature = wasmi::Signature::new(&[ValueType::I32][..], Some(ValueType::I32));
+        if alloc.signature() != &expected_signature {
+            anyhow::bail!(
+                "invalid signature for `alloc` export: {:?}, expected: {:?}",
+                alloc.signature(),
+                expected_signature
+            );
+        }
+
+        abi.instance = Some(instance.clone());
```

Comment:
Since now both `WasmState` and `WasmInterface` have `instance` as a field, maybe they could be merged into one struct like you had at the beginning?

Reply:
I tried but it was still a bit messy, but I agree we should look into it more now.